File: src/commands/query-metrics.ts
import { readFile } from 'fs/promises';
import chalk from 'chalk';
import type { Command } from 'commander';
import { join } from 'path';

interface StateHistoryEntry {
  state: string;
  timestamp: string;
}

interface WorkUnit {
  id: string;
  stateHistory?: StateHistoryEntry[];
  [key: string]: unknown;
}

interface WorkUnitsData {
  workUnits: Record<string, WorkUnit>;
  states: Record<string, string[]>;
}

interface MetricsResult {
  cycleTime?: string;
  timePerState?: Record<string, string>;
  aggregateMetrics?: {
    totalWorkUnits: number;
    completedWorkUnits: number;
    averageCycleTime?: string;
    byType?: Record<string, { count: number; averageCycleTime?: string }>;
  };
}

export async function queryMetrics(options: {
  workUnitId?: string;
  type?: 'story' | 'task' | 'bug';
  cwd?: string;
}): Promise<MetricsResult> {
  const cwd = options.cwd || process.cwd();
  const workUnitsFile = join(cwd, 'spec', 'work-units.json');

  try {
    // Read work units
    const content = await readFile(workUnitsFile, 'utf-8');
    const data: WorkUnitsData = JSON.parse(content);

    // Single work unit metrics
    if (options.workUnitId) {
      // Check if work unit exists
      if (!data.workUnits[options.workUnitId]) {
        throw new Error(`Work unit ${options.workUnitId} not found`);
      }

      const workUnit = data.workUnits[options.workUnitId];

      if (!workUnit.stateHistory || workUnit.stateHistory.length === 0) {
        throw new Error(`Work unit ${options.workUnitId} has no state history`);
      }

      // Calculate cycle time (first to last state)
      const firstTimestamp = new Date(workUnit.stateHistory[0].timestamp);
      const lastTimestamp = new Date(
        workUnit.stateHistory[workUnit.stateHistory.length - 1].timestamp
      );
      const cycleTimeMs = lastTimestamp.getTime() - firstTimestamp.getTime();
      const cycleTimeHours = Math.round(cycleTimeMs / (1000 * 60 * 60));

      // Calculate time per state
      const timePerState: Record<string, string> = {};
      for (let i = 0; i < workUnit.stateHistory.length - 1; i++) {
        const currentState = workUnit.stateHistory[i];
        const nextState = workUnit.stateHistory[i + 1];
        const currentTime = new Date(currentState.timestamp);
        const nextTime = new Date(nextState.timestamp);
        const durationMs = nextTime.getTime() - currentTime.getTime();
        const durationHours = Math.round(durationMs / (1000 * 60 * 60));
        timePerState[currentState.state] =
          `${durationHours} hour${durationHours !== 1 ? 's' : ''}`;
      }

      return {
        cycleTime: `${cycleTimeHours} hour${cycleTimeHours !== 1 ? 's' : ''}`,
        timePerState,
      };
    }

    // Aggregate metrics (when no specific work unit ID)
    let workUnits = Object.values(data.workUnits);

    // Filter by type if specified
    if (options.type) {
      workUnits = workUnits.filter(wu => {
        const type = wu.type || 'story';
        return type === options.type;
      });
    }

    const totalWorkUnits = workUnits.length;
    const completedWorkUnits = workUnits.filter(wu => wu.status === 'done')
      .length;

    // Calculate average cycle time for completed work units
    const completedWithHistory = workUnits.filter(
      wu => wu.status === 'done' && wu.stateHistory && wu.stateHistory.length > 0
    );

    let averageCycleTime: string | undefined;
    if (completedWithHistory.length > 0) {
      const totalCycleTimeMs = completedWithHistory.reduce((sum, wu) => {
        const firstTimestamp = new Date(wu.stateHistory![0].timestamp);
        const lastTimestamp = new Date(
          wu.stateHistory![wu.stateHistory!.length - 1].timestamp
        );
        return sum + (lastTimestamp.getTime() - firstTimestamp.getTime());
      }, 0);
      const avgHours = Math.round(
        totalCycleTimeMs / (1000 * 60 * 60) / completedWithHistory.length
      );
      averageCycleTime = `${avgHours} hour${avgHours !== 1 ? 's' : ''}`;
    }

    // Break down by type if no type filter specified
    let byType: Record<string, { count: number; averageCycleTime?: string }> | undefined;
    if (!options.type) {
      byType = {};
      const types = ['story', 'task', 'bug'] as const;
      for (const type of types) {
        const typeWorkUnits = workUnits.filter(wu => {
          const wuType = wu.type || 'story';
          return wuType === type;
        });
        const typeCompleted = typeWorkUnits.filter(
          wu =>
            wu.status === 'done' && wu.stateHistory && wu.stateHistory.length > 0
        );

        let typeAvgCycleTime: string | undefined;
        if (typeCompleted.length > 0) {
          const typeTotalCycleTimeMs = typeCompleted.reduce((sum, wu) => {
            const firstTimestamp = new Date(wu.stateHistory![0].timestamp);
            const lastTimestamp = new Date(
              wu.stateHistory![wu.stateHistory!.length - 1].timestamp
            );
            return sum + (lastTimestamp.getTime() - firstTimestamp.getTime());
          }, 0);
          const typeAvgHours = Math.round(
            typeTotalCycleTimeMs / (1000 * 60 * 60) / typeCompleted.length
          );
          typeAvgCycleTime = `${typeAvgHours} hour${typeAvgHours !== 1 ? 's' : ''}`;
        }

        byType[type] = {
          count: typeWorkUnits.length,
          averageCycleTime: typeAvgCycleTime,
        };
      }
    }

    return {
      aggregateMetrics: {
        totalWorkUnits,
        completedWorkUnits,
        averageCycleTime,
        byType,
      },
    };
  } catch (error: unknown) {
    if (error instanceof Error) {
      throw new Error(`Failed to query metrics: ${error.message}`);
    }
    throw error;
  }
}

export function registerQueryMetricsCommand(program: Command): void {
  program
    .command('query-metrics')
    .description('Query project metrics and statistics')
    .option('--work-unit-id <id>', 'Specific work unit to query metrics for')
    .option('--type <type>', 'Filter by work item type: story, task, or bug')
    .option('--format <format>', 'Output format: text or json', 'text')
    .action(
      async (options: {
        workUnitId?: string;
        type?: 'story' | 'task' | 'bug';
        format?: string;
      }) => {
        try {
          const result = await queryMetrics({
            workUnitId: options.workUnitId,
            type: options.type,
          });
          if (options.format === 'json') {
            console.log(JSON.stringify(result, null, 2));
          } else {
            // Text output for aggregate metrics
            if (result.aggregateMetrics) {
              console.log(chalk.bold('\nProject Metrics'));
              console.log('');
              console.log(
                `Total Work Units: ${result.aggregateMetrics.totalWorkUnits}`
              );
              console.log(
                `Completed Work Units: ${result.aggregateMetrics.completedWorkUnits}`
              );
              if (result.aggregateMetrics.averageCycleTime) {
                console.log(
                  `Average Cycle Time: ${result.aggregateMetrics.averageCycleTime}`
                );
              }
              if (result.aggregateMetrics.byType) {
                console.log('');
                console.log(chalk.bold('By Type:'));
                for (const [type, data] of Object.entries(
                  result.aggregateMetrics.byType
                )) {
                  console.log(
                    `  ${type}: ${data.count} work unit${data.count !== 1 ? 's' : ''}`
                  );
                  if (data.averageCycleTime) {
                    console.log(`    Average Cycle Time: ${data.averageCycleTime}`);
                  }
                }
              }
            } else if (result.cycleTime) {
              // Single work unit output
              console.log(chalk.bold('\nWork Unit Metrics'));
              console.log('');
              console.log(`Cycle Time: ${result.cycleTime}`);
              if (result.timePerState) {
                console.log('');
                console.log(chalk.bold('Time Per State:'));
                for (const [state, time] of Object.entries(result.timePerState)) {
                  console.log(`  ${state}: ${time}`);
                }
              }
            }
          }
        } catch (error: any) {
          console.error(chalk.red('✗ Query failed:'), error.message);
          process.exit(1);
        }
      }
    );
}

File: codelet/fspec-core/src/commands/query_metrics.rs
//! `query-metrics` — Rust port of `src/commands/query-metrics.ts` (RPC-261).
//!
//! Computes cycle-time and aggregate completion statistics from
//! `spec/work-units.json`. Two mutually-exclusive output shapes share the
//! same return type: when `workUnitId` is supplied we emit
//! `{ cycleTime, timePerState }` for that one unit; otherwise we emit
//! `{ aggregateMetrics: { totalWorkUnits, completedWorkUnits,
//! averageCycleTime?, byType? } }` across the entire workspace.
//!
//! Both invocation paths (the LLM-facing dispatcher AND the standalone
//! fspec Rust binary's clap subcommand) call this single function —
//! RPC-003 §7/§11 two-front-doors invariant.
//!
//! ## TS parity rules
//!
//! * **No auto-create** of `spec/work-units.json`. Missing file → error
//!   wrapped as `Failed to query metrics: <inner>`.
//! * **stateHistory is read from `wu.extra`** — see RPC-261 architecture
//!   decision. We deliberately do NOT extend the typed `WorkUnit` struct
//!   so other ported commands continue to round-trip the field
//!   transparently.
//! * **Aggregate `byType` key order**: literal `["story","task","bug"]` —
//!   matches the TS array iteration at `src/commands/query-metrics.ts:130`.
//! * **`wu.type || 'story'`** falsy-collapse via [`WorkUnit::type_str`].
//! * **Time-per-state**: walk `[0..len-2]`, store on the EARLIER state's
//!   key, hours rounded half-away-from-zero (TS `Math.round` semantics).
//! * **Error wrapping**: all errors emitted from this command MUST include
//!   the literal prefix `"Failed to query metrics: "` so the agent loop
//!   and CLI both see TS-equivalent diagnostics.

use std::path::Path;

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};

use crate::error::FspecCoreError;
use crate::io::io_error::format_io_error;
use crate::types::work_unit::{WorkUnit, WorkUnitsData};

// ─────────────────────────────────────────────────────────────────────────
// Arg shape (camelCase via serde)
// ─────────────────────────────────────────────────────────────────────────

#[derive(Debug, Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
struct QueryMetricsArgs {
    work_unit_id: Option<String>,
    r#type: Option<String>,
    format: Option<String>,
}

// ─────────────────────────────────────────────────────────────────────────
// Result shapes — explicit field declaration order for byte-stable JSON.
// `IndexMap` preserves insertion order for `byType` and `timePerState`
// regardless of whether the `preserve_order` feature is enabled on
// `serde_json`, because `indexmap`'s own `Serialize` impl iterates by
// insertion.
// ─────────────────────────────────────────────────────────────────────────

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct SingleResult {
    cycle_time: String,
    time_per_state: IndexMap<String, String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct AggregateResult {
    aggregate_metrics: AggregateMetrics,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct AggregateMetrics {
    total_work_units: usize,
    completed_work_units: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    average_cycle_time: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    by_type: Option<IndexMap<String, TypeBreakdown>>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct TypeBreakdown {
    count: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    average_cycle_time: Option<String>,
}

/// Internal discriminated output — emitted in either branch and then
/// serialised to JSON/text in [`run`].
enum MetricsOutput {
    Single(SingleResult),
    Aggregate(AggregateResult),
}

// ─────────────────────────────────────────────────────────────────────────
// Public entry point
// ─────────────────────────────────────────────────────────────────────────

pub async fn run(args_json: &str, project_root: &Path) -> Result<String, FspecCoreError> {
    let args: QueryMetricsArgs =
        serde_json::from_str(args_json).map_err(|e| FspecCoreError::InvalidArgs {
            command: "query-metrics",
            reason: format!("failed to parse args: {e}"),
        })?;

    let data = load_work_units(project_root)?;

    let output = compute(&data, args.work_unit_id.as_deref(), args.r#type.as_deref())?;

    match args.format.as_deref() {
        Some("text") => Ok(render_text(&output)),
        // Default to JSON for parity with the dispatcher contract. The CLI
        // bridge always passes an explicit `format`, so the only case where
        // we fall through here is a dispatcher call that omits `format`.
        _ => to_json(&output),
    }
}

fn to_json(output: &MetricsOutput) -> Result<String, FspecCoreError> {
    let err = |e: serde_json::Error| FspecCoreError::InvalidArgs {
        command: "query-metrics",
        reason: wrap(format!("failed to serialize result: {e}")),
    };
    match output {
        MetricsOutput::Single(s) => serde_json::to_string_pretty(s).map_err(err),
        MetricsOutput::Aggregate(a) => serde_json::to_string_pretty(a).map_err(err),
    }
}

// ─────────────────────────────────────────────────────────────────────────
// I/O — no auto-create; ALL errors wrapped with the canonical prefix.
// ─────────────────────────────────────────────────────────────────────────

fn load_work_units(project_root: &Path) -> Result<WorkUnitsData, FspecCoreError> {
    let path = project_root.join("spec").join("work-units.json");
    let raw = std::fs::read_to_string(&path).map_err(|e| FspecCoreError::InvalidArgs {
        command: "query-metrics",
        reason: wrap(format_io_error(&e, &path.display().to_string())),
    })?;
    serde_json::from_str(&raw).map_err(|e| FspecCoreError::InvalidArgs {
        command: "query-metrics",
        reason: wrap(format!("Unexpected token in JSON: {e}")),
    })
}

fn wrap(msg: impl Into<String>) -> String {
    format!("Failed to query metrics: {}", msg.into())
}

// ─────────────────────────────────────────────────────────────────────────
// Compute branches
// ─────────────────────────────────────────────────────────────────────────

fn compute(
    data: &WorkUnitsData,
    work_unit_id: Option<&str>,
    type_filter: Option<&str>,
) -> Result<MetricsOutput, FspecCoreError> {
    if let Some(id) = work_unit_id {
        return compute_single(data, id);
    }
    Ok(MetricsOutput::Aggregate(compute_aggregate(
        data,
        type_filter,
    )))
}

fn compute_single(data: &WorkUnitsData, id: &str) -> Result<MetricsOutput, FspecCoreError> {
    let wu = data
        .work_units
        .get(id)
        .ok_or_else(|| FspecCoreError::InvalidArgs {
            command: "query-metrics",
            reason: wrap(format!("Work unit {id} not found")),
        })?;

    let history = extract_history(wu);
    if history.is_empty() {
        return Err(FspecCoreError::InvalidArgs {
            command: "query-metrics",
            reason: wrap(format!("Work unit {id} has no state history")),
        });
    }

    let first_ms = parse_iso_to_ms(&history[0].timestamp);
    let last_ms = parse_iso_to_ms(&history[history.len() - 1].timestamp);
    let cycle_hours = round_hours(last_ms - first_ms);

    let mut time_per_state: IndexMap<String, String> = IndexMap::new();
    for i in 0..history.len().saturating_sub(1) {
        let cur = &history[i];
        let next = &history[i + 1];
        let dur_ms = parse_iso_to_ms(&next.timestamp) - parse_iso_to_ms(&cur.timestamp);
        let dur_hours = round_hours(dur_ms);
        time_per_state.insert(cur.state.clone(), format_hours(dur_hours));
    }

    Ok(MetricsOutput::Single(SingleResult {
        cycle_time: format_hours(cycle_hours),
        time_per_state,
    }))
}

fn compute_aggregate(data: &WorkUnitsData, type_filter: Option<&str>) -> AggregateResult {
    // TS: `Object.values(data.workUnits)` — insertion order via IndexMap.
    let all: Vec<&WorkUnit> = data.work_units.values().collect();

    // TS: filter by `wu.type || 'story'` only when --type is supplied.
    let filtered: Vec<&WorkUnit> = match type_filter {
        Some(t) => all.into_iter().filter(|wu| wu.type_str() == t).collect(),
        None => all,
    };

    let total = filtered.len();
    let completed = filtered
        .iter()
        .filter(|wu| wu.status.as_str() == "done")
        .count();

    // TS: average over `status === 'done' && stateHistory.length > 0`.
    let completed_with_history: Vec<&WorkUnit> = filtered
        .iter()
        .copied()
        .filter(|wu| wu.status.as_str() == "done" && !extract_history(wu).is_empty())
        .collect();
    let average_cycle_time = avg_cycle(&completed_with_history);

    // TS: byType populated ONLY when --type is NOT supplied.
    let by_type = if type_filter.is_none() {
        let mut m: IndexMap<String, TypeBreakdown> = IndexMap::new();
        // Literal TS canonical order: story, task, bug.
        for t in ["story", "task", "bug"] {
            let type_units: Vec<&WorkUnit> = filtered
                .iter()
                .copied()
                .filter(|wu| wu.type_str() == t)
                .collect();
            let type_completed: Vec<&WorkUnit> = type_units
                .iter()
                .copied()
                .filter(|wu| wu.status.as_str() == "done" && !extract_history(wu).is_empty())
                .collect();
            let avg = avg_cycle(&type_completed);
            m.insert(
                t.to_string(),
                TypeBreakdown {
                    count: type_units.len(),
                    average_cycle_time: avg,
                },
            );
        }
        Some(m)
    } else {
        None
    };

    AggregateResult {
        aggregate_metrics: AggregateMetrics {
            total_work_units: total,
            completed_work_units: completed,
            average_cycle_time,
            by_type,
        },
    }
}

/// Average cycle time across a slice of done-with-history units. Returns
/// `None` when the slice is empty (TS leaves `averageCycleTime`
/// `undefined`, which JSON-stringify omits).
fn avg_cycle(units: &[&WorkUnit]) -> Option<String> {
    if units.is_empty() {
        return None;
    }
    let total_ms: i64 = units
        .iter()
        .map(|wu| {
            let h = extract_history(wu);
            let first = parse_iso_to_ms(&h[0].timestamp);
            let last = parse_iso_to_ms(&h[h.len() - 1].timestamp);
            last - first
        })
        .sum();
    // TS: Math.round(totalCycleTimeMs / 3_600_000 / completedCount)
    let avg_hours = ((total_ms as f64) / 3_600_000.0 / units.len() as f64).round() as i64;
    Some(format_hours(avg_hours))
}

// ─────────────────────────────────────────────────────────────────────────
// stateHistory extraction (from `wu.extra`, per RPC-261 design)
// ─────────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
struct HistoryEntry {
    state: String,
    timestamp: String,
}

fn extract_history(wu: &WorkUnit) -> Vec<HistoryEntry> {
    wu.extra
        .get("stateHistory")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| {
                    let o = v.as_object()?;
                    let state = o.get("state")?.as_str()?.to_string();
                    let timestamp = o.get("timestamp")?.as_str()?.to_string();
                    Some(HistoryEntry { state, timestamp })
                })
                .collect()
        })
        .unwrap_or_default()
}

// ─────────────────────────────────────────────────────────────────────────
// Time helpers — minimal RFC-3339 ms parser + half-away-from-zero round.
// Mirrors the `parse_iso_to_ms` / `days_from_civil` helpers in
// `query_work_units.rs`; we keep a private copy here so this command stays
// independent of that module's internal API surface.
// ─────────────────────────────────────────────────────────────────────────

fn format_hours(h: i64) -> String {
    if h == 1 || h == -1 {
        format!("{h} hour")
    } else {
        format!("{h} hours")
    }
}

/// TS `Math.round(ms / 3_600_000)` — half-away-from-zero rounding.
/// `f64::round()` is half-away-from-zero, which matches TS for both
/// positive and negative deltas (TS's `Math.round` rounds half UP, so
/// `Math.round(-0.5) === 0` whereas Rust gives `-1.0` — we accept this
/// tiny corner-case divergence for negative deltas which never appear in
/// real fspec data; documented here for completeness).
fn round_hours(ms: i64) -> i64 {
    ((ms as f64) / 3_600_000.0).round() as i64
}

fn parse_iso_to_ms(s: &str) -> i64 {
    let trimmed = s.trim_end_matches('Z');
    let (date_part, time_part) = match trimmed.split_once('T') {
        Some((d, t)) => (d, t),
        None => return 0,
    };
    let date_iter: Vec<&str> = date_part.split('-').collect();
    if date_iter.len() != 3 {
        return 0;
    }
    let year: i64 = date_iter[0].parse().unwrap_or(0);
    let month: u32 = date_iter[1].parse().unwrap_or(0);
    let day: u32 = date_iter[2].parse().unwrap_or(0);
    let time_core = time_part.split('.').next().unwrap_or("");
    let t_iter: Vec<&str> = time_core.split(':').collect();
    if t_iter.len() != 3 {
        return 0;
    }
    let hh: i64 = t_iter[0].parse().unwrap_or(0);
    let mm: i64 = t_iter[1].parse().unwrap_or(0);
    let ss: i64 = t_iter[2].parse().unwrap_or(0);
    let frac_ms: i64 = time_part
        .split('.')
        .nth(1)
        .and_then(|f| f.parse::<i64>().ok())
        .unwrap_or(0);
    let days = days_from_civil(year as i32, month, day);
    let secs = days * 86_400 + hh * 3_600 + mm * 60 + ss;
    secs * 1_000 + frac_ms
}

fn days_from_civil(y: i32, m: u32, d: u32) -> i64 {
    let y = if m <= 2 { y - 1 } else { y } as i64;
    let era = if y >= 0 { y / 400 } else { (y - 399) / 400 };
    let yoe = y - era * 400;
    let m = m as i64;
    let d = d as i64;
    let doy = (153 * (if m > 2 { m - 3 } else { m + 9 }) + 2) / 5 + d - 1;
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
    era * 146_097 + doe - 719_468
}

// ─────────────────────────────────────────────────────────────────────────
// Text rendering — parity with TS `output.log(...)` sequence at
// src/commands/query-metrics.ts:202-247.
// ─────────────────────────────────────────────────────────────────────────

fn render_text(output: &MetricsOutput) -> String {
    let mut s = String::new();
    match output {
        MetricsOutput::Aggregate(a) => {
            s.push('\n');
            s.push_str("Project Metrics\n");
            s.push('\n');
            s.push_str(&format!(
                "Total Work Units: {}\n",
                a.aggregate_metrics.total_work_units
            ));
            s.push_str(&format!(
                "Completed Work Units: {}\n",
                a.aggregate_metrics.completed_work_units
            ));
            if let Some(avg) = &a.aggregate_metrics.average_cycle_time {
                s.push_str(&format!("Average Cycle Time: {avg}\n"));
            }
            if let Some(by_type) = &a.aggregate_metrics.by_type {
                s.push('\n');
                s.push_str("By Type:\n");
                for (t, b) in by_type {
                    let unit_word = if b.count == 1 { "unit" } else { "units" };
                    s.push_str(&format!("  {t}: {} work {unit_word}\n", b.count));
                    if let Some(avg) = &b.average_cycle_time {
                        s.push_str(&format!("    Average Cycle Time: {avg}\n"));
                    }
                }
            }
        }
        MetricsOutput::Single(single) => {
            s.push('\n');
            s.push_str("Work Unit Metrics\n");
            s.push('\n');
            s.push_str(&format!("Cycle Time: {}\n", single.cycle_time));
            // TS always emits the heading when timePerState is truthy
            // (it always is for the single-unit branch — even an empty
            // object passes the `if (result.timePerState)` check).
            s.push('\n');
            s.push_str("Time Per State:\n");
            for (state, time) in &single.time_per_state {
                s.push_str(&format!("  {state}: {time}\n"));
            }
        }
    }
    s
}

#[cfg(test)]
mod tests {
    #![allow(
        clippy::unwrap_used,
        clippy::expect_used,
        clippy::panic,
        clippy::useless_vec
    )]
    use super::*;

    #[test]
    fn format_hours_pluralises_around_one() {
        assert_eq!(format_hours(0), "0 hours");
        assert_eq!(format_hours(1), "1 hour");
        assert_eq!(format_hours(-1), "-1 hour");
        assert_eq!(format_hours(2), "2 hours");
    }

    #[test]
    fn round_hours_half_away_from_zero() {
        // 1.5h → 2h
        assert_eq!(round_hours(5_400_000), 2);
        // 1.4h → 1h
        assert_eq!(round_hours(5_040_000), 1);
        // 0.5h → 1h (half-away-from-zero)
        assert_eq!(round_hours(1_800_000), 1);
    }

    #[test]
    fn parse_iso_to_ms_round_trips_simple_utc_stamps() {
        // 2026-01-01T00:00:00.000Z vs 2026-01-01T05:00:00.000Z = 5h delta
        let a = parse_iso_to_ms("2026-01-01T00:00:00.000Z");
        let b = parse_iso_to_ms("2026-01-01T05:00:00.000Z");
        assert_eq!(b - a, 5 * 3_600_000);
    }

    #[test]
    fn wrap_prepends_canonical_prefix() {
        assert_eq!(
            wrap("Work unit X not found"),
            "Failed to query metrics: Work unit X not found"
        );
    }
}
